core/src/main/scala/org/apache/spark/sql/delta/commands/MergeIntoCommand.scala

```diff
@@ -583,16 +604,24 @@ case class MergeIntoCommand(
     // with value `true`, one to each side of the join. Whether this field is null or not after
     // the outer join, will allow us to identify whether the resultant joined row was a
     // matched inner result or an unmatched result with null on one side.
+    // We add row IDs to the targetDF if we have a delete-when-matched clause with duplicate
+    // matches and CDC is enabled, and additionally add row IDs to the source if we also have an
+    // insert clause. See above at isDeleteWithDuplicateMatchesAndCdc definition for more details.
     val joinedDF = {
-      val sourceDF = Dataset.ofRows(spark, source)
+      var sourceDF = Dataset.ofRows(spark, source)
```

Comment:
this is getting very nested. i think it will be better to pull out the sourceDF and targetDF outside the { } of joinedDF